bodhi/server/models.py

```diff
@@ -1555,7 +1555,8 @@ def get_changelog(self, timelimit=0):
 
 class Update(Base):
     """
-    This model represents an update.
+    Update model. Single update contains builds for one or more packages, each
+    package can have more than one build.
```

Comment:
I would word it a bit differently. I'd probably keep the original text and add a new paragraph to expand on it:

```
This model represents an update.

An update contains a set of Build objects that the packager is proposing to update atomically in the release.
```

Though a package can have more than one build in the database, it cannot have more than one build in the same update.

Reply:
I don't get it. If package can not have more than one build, then why the documentation says that update contains a set of builds?

Reply:
A package can have more than one build, but an update can only have one build per package.

Reply:
In other words the update is a just a label or tag on a package build. Is that right?

Reply:
There is contradiction here. First it is said that "update contains a set of Build objects" but then you say that update can only have one build. Where is the truth?

Reply:
@abitrolly The update has a set of builds. Each build has an associated package. Bodhi enforces a rule that there can only be one build per package associated with an update. As an example of what I mean, it is OK for an update to have httpd-1.0.0-1.fc29 and python3-3.7.0-1.fc29 in it. It is not allowed for a single update to have both httpd-1.0.0-1.fc29 and httpd-1.0.0-2.fc29 in it.

Reply:
In human world, update contains new package versions. Then it becomes logical. Those builds are also packages, right?

Reply:
There is a bug in semantic model, which means that `bodhi` interface is confusing. Here is the https://bodhi.fedoraproject.org/updates/?releases=F30&status=pending

![image](https://user-images.githubusercontent.com/8781107/57277104-1cb7bc80-70ac-11e9-9501-db0419e8f4cc.png)

As you may see `Updates` list actually lists `packages`, not updates. It lists ` oniguruma-6.9.2-1.fc30` package name and not update name, which is `FEDORA-2019-abe79f6459`. Then following https://bodhi.fedoraproject.org/updates/FEDORA-2019-abe79f6459 there is no mention of `oniguruma-6.9.2-1.fc30` unless somebody goes to `Builds` tab.

In the end, the `new package` is called `update` then `build`, and that's why I couldn't understand the code.

How to find an update for more than one build to see how it looks like?

Reply:
Found an `update pack`. It doesn't stand out from the UI except for the long name, which is not completely evident.

![image](https://user-images.githubusercontent.com/8781107/57277699-5b01ab80-70ad-11e9-9b36-276a12691ab9.png)

There are 5 builds and an instruction how to install it. https://bodhi.fedoraproject.org/updates/FEDORA-2019-7ee81eedb3

These `update packs` could stand out to make it clear that `update` may contain a bunch of packages. A package count on update could do the trick.

Reply:
> Those builds are also packages, right?

Unfortunately, it's not that simple. In the most common case, the "builds" will be RPM packages, but they could e.g. also be containers or modules. For example, the [FEDORA-MODULAR-2019-088c6693d6](https://bodhi.fedoraproject.org/updates/FEDORA-MODULAR-2019-088c6693d6) update contains the [gimp-2.10-2920190409123811.6c81f848](https://koji.fedoraproject.org/koji/buildinfo?buildID=1247188) "build" for stream 2.10 of the gimp module which in turn contains the [gimp-2.10.10-1.module_f30+3878+d4400ad3](https://koji.fedoraproject.org/koji/buildinfo?buildID=1247146) package as an installable artifact.

Reply:
There is no way one can figure this out from reading the source code. Without simplifying this part or documenting it, there is a little chance for potential contributors to understand.